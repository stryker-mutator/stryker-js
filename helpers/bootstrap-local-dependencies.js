// @ts-check
import { LocalInstaller, progress } from 'install-local';
import fs from 'fs';
import glob from 'glob';
import path from 'path';

console.log('starting installation of local dependencies');

/**
 * @param {string} pattern 
 * @param {glob.IOptions} options 
 * @returns {Promise<string[]>}
 */
function globAsPromised(pattern, options) {
  return new Promise((res, rej) => {
    glob(pattern, options, (err, matches) => {
      if (err) {
        rej(err);
      } else {
        res(matches);
      }
    });
  });
}

/**
 * @typedef Package
 * @property {{ [name: string]: string }} localDependencies

/**
 * Installs local dependencies in one go,
 * reads the package.json files and installs the local dependencies marked there.
 * The packageJsonGlobs default to ./package.json and test/* /package.json files 
 * @param {string} directory
 * @param {string[]} packageJsonGlobs
 * @returns {Promise<void>}
 */
export async function bootstrapLocalDependencies(directory, packageJsonGlobs = ['package.json', 'test/*/package.json']) {
  console.log(`bootstrap ${path.resolve(directory)} (using ${packageJsonGlobs.join(',')})`);
  const files = (await Promise.all(packageJsonGlobs.map((globPattern) => globAsPromised(globPattern, { cwd: path.resolve(directory) })))).flat();
  const packages = await Promise.all(files.map(fileName => fs.promises.readFile(fileName, 'utf-8')
    .then((rawContent) => {
      /**
       * @type {Package}
       */
      const content = JSON.parse(rawContent);
      return { dir: path.dirname(fileName), content };
    })));
  /**
   * @type {import('install-local').ListByPackage}
   */
  const sourcesByTarget = {};
  for (const pkg of packages) {
    const localDeps = pkg.content.localDependencies;
    if (localDeps) {
      sourcesByTarget[path.resolve(pkg.dir)] = Object.keys(localDeps).map(key => path.resolve(pkg.dir, localDeps[key]));
    }
  }
  const localInstaller = new LocalInstaller(sourcesByTarget);
  progress(localInstaller);
  await localInstaller.install();
}
