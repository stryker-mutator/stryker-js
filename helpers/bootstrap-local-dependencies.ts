import { ListByPackage, LocalInstaller, progress } from 'install-local';
import * as fs from 'fs';
import glob = require('glob');
import path = require('path');

console.log('starting installation of local dependencies');

function globAsPromised(pattern: string, options: glob.IOptions) {
  return new Promise<string[]>((res, rej) => {
    glob(pattern, options, (err, matches) => {
      if (err) {
        rej(err);
      } else {
        res(matches);
      }
    });
  });
}

function readFile(fileName: string) {
  return new Promise<string>((res, rej) => fs.readFile(fileName, 'utf8', (err, content) => {
    if (err) {
      rej(err);
    } else {
      res(content);
    }
  }));
}

interface Package {
  localDependencies?: { [name: string]: string };
}

/**
 * Installs local dependencies in one go,
 * reads package.json and test/* /package.json files and installs the local dependencies marked there
 * @param directory the directory where the tests live
 */
export async function bootstrapLocalDependencies(directory: string) {
  console.log('bootstrap ' + path.resolve(directory));
  const files = await globAsPromised('{package.json,test/*/package.json}', { cwd: path.resolve(directory) });
  const packages = await Promise.all(files.map(fileName => readFile(fileName)
    .then(content => ({ dir: path.dirname(fileName), content: JSON.parse(content) as Package }))));
  const sourcesByTarget: ListByPackage = {};
  for (const pkg of packages) {
    const localDeps = pkg.content.localDependencies;
    if (localDeps) {
      sourcesByTarget[path.resolve(pkg.dir)] = Object.keys(localDeps).map(key => path.resolve(pkg.dir, localDeps[key]));
    }
  }
  const localInstaller = new LocalInstaller(sourcesByTarget);
  progress(localInstaller);
  await localInstaller.install();
}
