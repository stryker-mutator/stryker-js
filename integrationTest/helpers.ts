import * as path from 'path';
import * as fs from 'mz/fs';
import { expect } from 'chai';
import { ScoreResult } from 'stryker-api/report';

export async function readScoreResult(eventResultDirectory = path.resolve('reports', 'mutation', 'events')) {
    const allReportFiles = await fs.readdir(eventResultDirectory);
    const scoreResultReportFile = allReportFiles.find(file => !!file.match(/.*onScoreCalculated.*/));
    expect(scoreResultReportFile).ok;
    const scoreResultContent = await fs.readFile(path.resolve(eventResultDirectory, scoreResultReportFile || ''), 'utf8');
    return JSON.parse(scoreResultContent) as ScoreResult;
}