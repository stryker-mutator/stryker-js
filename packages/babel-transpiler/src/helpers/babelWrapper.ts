/**
 * This wrapper is needed because babel/core only exports
 * non-stubbable es6 properties. See node_modules/@babel/core/lib/index.js
 */
export * from '@babel/core';
