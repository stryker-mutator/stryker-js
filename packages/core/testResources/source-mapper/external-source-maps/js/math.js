"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function max() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return Math.max.apply(Math, numbers);
}
exports.max = max;
function total() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return numbers.reduce(function (a, b) { return a + b; });
}
exports.total = total;
//# sourceMappingURL=math.js.map