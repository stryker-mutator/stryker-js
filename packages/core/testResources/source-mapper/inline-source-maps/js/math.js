"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function max() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return Math.max.apply(Math, numbers);
}
exports.max = max;
function total() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return numbers.reduce(function (a, b) { return a + b; });
}
exports.total = total;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoibWF0aC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uL3RzL3NyYy9tYXRoLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBRUE7SUFBb0IsaUJBQW9CO1NBQXBCLFVBQW9CLEVBQXBCLHFCQUFvQixFQUFwQixJQUFvQjtRQUFwQiw0QkFBb0I7O0lBQ3RDLE1BQU0sQ0FBQyxJQUFJLENBQUMsR0FBRyxPQUFSLElBQUksRUFBUSxPQUFPLEVBQUU7QUFDOUIsQ0FBQztBQUZELGtCQUVDO0FBRUQ7SUFBc0IsaUJBQW9CO1NBQXBCLFVBQW9CLEVBQXBCLHFCQUFvQixFQUFwQixJQUFvQjtRQUFwQiw0QkFBb0I7O0lBQ3hDLE1BQU0sQ0FBQyxPQUFPLENBQUMsTUFBTSxDQUFDLFVBQUMsQ0FBQyxFQUFFLENBQUMsSUFBSyxPQUFBLENBQUMsR0FBRyxDQUFDLEVBQUwsQ0FBSyxDQUFDLENBQUM7QUFDekMsQ0FBQztBQUZELHNCQUVDIn0=