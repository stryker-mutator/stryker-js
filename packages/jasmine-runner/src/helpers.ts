import { TestResult, TestStatus } from '@stryker-mutator/api/test-runner';
import jasmine from 'jasmine';

export const helpers = {
  toStrykerTestResult(specResult: jasmine.SpecResult, timeSpentMs: number): TestResult {
    const baseResult = {
      id: specResult.id.toString(),
      name: specResult.fullName,
      timeSpentMs,
    };
    if (specResult.status === 'disabled' || specResult.status === 'pending' || specResult.status === 'excluded') {
      return {
        ...baseResult,
        status: TestStatus.Skipped,
      };
    } else if (!specResult.failedExpectations || specResult.failedExpectations.length === 0) {
      return {
        ...baseResult,
        status: TestStatus.Success,
      };
    } else {
      return {
        ...baseResult,
        status: TestStatus.Failed,
        failureMessage: specResult.failedExpectations.map((failedExpectation) => failedExpectation.message).join(','),
      };
    }
  },
  createJasmine(options: jasmine.JasmineOptions): jasmine {
    return new jasmine(options);
  },
};
