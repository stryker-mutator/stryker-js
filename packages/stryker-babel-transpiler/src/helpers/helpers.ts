import * as path from 'path';

export function toJSFileName(fileName: string) {
  const fileNameInfo = path.parse(fileName);
  const newFileName = path.join(fileNameInfo.dir, fileNameInfo.name + '.js');
  return newFileName;
}
