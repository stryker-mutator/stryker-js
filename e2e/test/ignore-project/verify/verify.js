import { expect } from 'chai';
import { MutantStatus } from 'mutation-testing-report-schema';

import { expectMetricsJsonToMatchSnapshot, readMutationTestingJsonResultAsMetricsResult } from '../../../helpers.js';

describe('After running stryker on jest-react project', () => {
  it('should report expected scores', async () => {
    await expectMetricsJsonToMatchSnapshot();
  });
  it('should report mutants that are disabled by a comment with correct ignore reason', async () => {
    const actualMetricsResult = await readMutationTestingJsonResultAsMetricsResult();
    const addResult = actualMetricsResult.systemUnderTestMetrics.childResults.find((file) => file.name.endsWith('Add.js')).file;
    const mutantsAtLine31 = addResult.mutants.filter(({ location }) => location.start.line === 31);
    const booleanLiteralMutants = mutantsAtLine31.filter(({ mutatorName }) => mutatorName === 'BooleanLiteral');
    const conditionalExpressionMutants = mutantsAtLine31.filter(({ mutatorName }) => mutatorName === 'ConditionalExpression');
    const equalityOperatorMutants = mutantsAtLine31.filter(({ mutatorName }) => mutatorName === 'EqualityOperator');
    booleanLiteralMutants.forEach((booleanMutant) => {
      expect(booleanMutant.status).eq(MutantStatus.Ignored);
      expect(booleanMutant.statusReason).eq('Ignore boolean and conditions');
    });
    conditionalExpressionMutants.forEach((conditionalMutant) => {
      expect(conditionalMutant.status).eq(MutantStatus.Ignored);
      expect(conditionalMutant.statusReason).eq('Ignore boolean and conditions');
    });
    equalityOperatorMutants.forEach((equalityMutant) => {
      expect(equalityMutant.status).eq(MutantStatus.NoCoverage);
    });
  });
  it('should report mutants that result from excluded mutators with the correct ignore reason', async () => {
    const actualMetricsResult = await readMutationTestingJsonResultAsMetricsResult();
    const circleResult = actualMetricsResult.systemUnderTestMetrics.childResults.find((file) => file.name.endsWith('Circle.js')).file;
    const mutantsAtLine3 = circleResult.mutants.filter(({ location }) => location.start.line === 3);
    mutantsAtLine3.forEach((mutant) => {
      expect(mutant.status).eq(MutantStatus.Ignored);
      expect(mutant.statusReason).eq('Ignored because of excluded mutation "ArithmeticOperator"');
    });
  });
});
