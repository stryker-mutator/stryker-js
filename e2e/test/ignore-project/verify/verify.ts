import { expect } from 'chai';
import { MutantStatus } from 'mutation-testing-report-schema';
import { expectMetricsJson, readMutationTestingJsonResult  } from '../../../helpers';

describe('After running stryker on jest-react project', () => {
  it('should report expected scores', async () => {
    await expectMetricsJson({
      killed: 8,
      ignored: 28,
      mutationScore: 53.33,
    });
  });

  /*
  -----------|---------|----------|-----------|------------|----------|---------|
  File       | % score | # killed | # timeout | # survived | # no cov | # error |
  -----------|---------|----------|-----------|------------|----------|---------|
  All files  |   53.33 |        8 |         0 |          0 |        7 |       0 |*/


  it('should report mutants that are disabled by a comment with correct ignore reason', async () => {
    const actualMetricsResult = await readMutationTestingJsonResult();
    const addResult = actualMetricsResult.systemUnderTestMetrics.childResults.find(file => file.name.endsWith('Add.js')).file!;
    const mutantsAtLine31 =  addResult.mutants.filter(({ location }) => location.start.line === 31)
    const booleanLiteralMutants = mutantsAtLine31.filter(({mutatorName}) => mutatorName === 'BooleanLiteral');
    const conditionalExpressionMutants = mutantsAtLine31.filter(({mutatorName}) => mutatorName === 'ConditionalExpression');
    const equalityOperatorMutants = mutantsAtLine31.filter(({mutatorName}) => mutatorName === 'EqualityOperator');
    booleanLiteralMutants.forEach((booleanMutant) => {
      expect(booleanMutant.status).eq(MutantStatus.Ignored);
      expect(booleanMutant.statusReason).eq('Ignore boolean and conditions');
    });
    conditionalExpressionMutants.forEach((conditionalMutant) => {
      expect(conditionalMutant.status).eq(MutantStatus.Ignored);
      expect(conditionalMutant.statusReason).eq('Ignore boolean and conditions');
    });

    equalityOperatorMutants.forEach((equalityMutant) => {
      expect(equalityMutant.status).eq(MutantStatus.NoCoverage);
    });
  });

  it('should report mutants that result from excluded mutators with the correct ignore reason', async () => {
    const actualMetricsResult = await readMutationTestingJsonResult();
    const circleResult = actualMetricsResult.systemUnderTestMetrics.childResults.find(file => file.name.endsWith('Circle.js')).file!;
    const mutantsAtLine3 =  circleResult.mutants.filter(({ location }) => location.start.line === 3)

    mutantsAtLine3.forEach((mutant) => {
      expect(mutant.status).eq(MutantStatus.Ignored);
      expect(mutant.statusReason).eq('Ignored because of excluded mutation "ArithmeticOperator"');
    });
  });
});
