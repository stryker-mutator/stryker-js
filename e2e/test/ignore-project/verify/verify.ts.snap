// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`After running stryker on jest-react project should report expected scores 1`] = `
Object {
  "compileErrors": 0,
  "ignored": 28,
  "killed": 8,
  "mutationScore": 53.333333333333336,
  "mutationScoreBasedOnCoveredCode": 100,
  "noCoverage": 7,
  "runtimeErrors": 0,
  "survived": 0,
  "timeout": 0,
  "totalCovered": 8,
  "totalDetected": 8,
  "totalInvalid": 0,
  "totalMutants": 43,
  "totalUndetected": 7,
  "totalValid": 15,
}
`;
